### PR TITLE
Attach generated default method to mapped trait

### LINKED ISSUE #1081: _default method doesn't set the default value of a mapped trait
Current behaviour:
``` python
class Person(HasTraits):

    married = Map({'yes': 1, 'no': 0})

    def _married_default(self):
        return 'no'


bob = Person()
print(bob.married)  # prints 'no'
print(bob.married_)  # prints <undefined>, but should print 0
```

The problem seems to lie in the way `_default` methods are attached to traits. This happens after all trait definitions have been processed and definitions for mapped traits have been created.
https://github.com/enthought/traits/blob/c3af7affed2d69b1e0239bc9e836f220fa6606ad/traits/has_traits.py#L621-L633
The code iterates over all traits and checks for their default methods. In the case of the example, it checks and finds `_married_default` which is then set as the default value of `married` trait. However, because there is no ` _married__default`, no default value is set for `married_` trait.

One solution I can think of is that if the trait name ends in `_`, which indicates a mapped trait, the code also looks for a default method using `name[:-1]` and sets the default value of the mapped trait to be a mapped version of the default method it found.

A sidenote: default value assignment through trait definition sets the default value of the mapped trait as expected via https://github.com/enthought/traits/blob/c3af7affed2d69b1e0239bc9e836f220fa6606ad/traits/trait_converters.py#L112-L125

### PR DESCRIPTION
**Pull request by ievacerny**
Closes #1081

~The mapped trait now gets a new generated default method _IF_ the key trait also has a default method. If the user defines `_default__trait` method, it is ignored to keep both mapped and key traits synced.~

EDIT: The PR now follows the approach detailed in a comment below: https://github.com/enthought/traits/pull/1091#issuecomment-630056494

**Checklist**
- [x] Tests
- ~[ ] Update API reference (`docs/source/traits_api_reference`)~
- ~[ ] Update User manual (`docs/source/traits_user_manual`)~
- ~[ ] Update type annotation hints in `traits-stubs`~

### CODE REVIEW

**Comment by ievacerny** [comment]
Regarding @kitchoi suggestion on the issue:
> Maybe you are already thinking doing the same, though I am not sure from the conversation above: Would it be possible put your default function in #1081 (comment) in this mapped_trait_for? It will need the trait name, but it seems that all uses of mapped_trait_for are in the context where the trait name is available already, so the function mapped_trait_for can be expanded to accept name as an argument.

The ugly part of the code is checking if the default method should be attached to the mapped trait, which can't be moved to `mapped_trait_for`. But if it's fine to attach a default method to all mapped traits, then the code could definitely be simplified. An issue with this would be attempting to map an `undefined` key trait (if it doesn't have a default value).

File: traits/has_traits.py [hunk 1]
```diff
@@ -305,6 +306,16 @@ def _property_method(class_dict, name):
     return class_dict.get(name)
 
 
+def _mapped_trait_default_method(trait, trait_name):
+
```
Comment (mdickinson) [inline traits/has_traits.py:310]: Please could we have a docstring here? (Even though it's a private function, it'll still help people who are reading and understanding the code.)

File: traits/has_traits.py [hunk 2]
```diff
@@ -305,6 +306,16 @@ def _property_method(class_dict, name):
...
+    def default_method(key_trait, key_trait_name, object):
+        return key_trait.handler.mapped_value(
+            getattr(object, key_trait_name, Undefined)
+        )
+
+    return partial(default_method, trait, trait_name)
```
Comment (mdickinson) [inline traits/has_traits.py:316]: I'm assuming that the `partial` use was necessary to make the default method is pickleable. Is that right?

File: traits/has_traits.py [hunk 3]
```diff
@@ -619,7 +630,23 @@ def update_traits_class_dict(class_name, bases, class_dict):
                 )
 
         handlers = [h for h in handlers if h is not None]
+
         default = _get_def(class_name, class_dict, [], "_%s_default" % name)
+        # In case of mapped traits, need to form a new default method
+        if (name[-1:] == '_'):
```
Comment (mdickinson) [inline traits/has_traits.py:636]: I guess the solution I was thinking of was different: that is, in the branch `if default is not None` below, set both the default value of the main trait _and_ (in the case of a mapped trait) the default value of the shadow trait at the same time. Would that work, or is there something I'm missing?

I'm not too worried about what happens if the user's created a `default` method just for the shadow trait; that seems as though it should almost always be an error (especially if there's _also_ a default for the main trait).

Comment (mdickinson) [reply traits/has_traits.py:636]: (Nitpick: unnecessary parentheses around the `if` condition.)

**Comment by kitchoi** [comment]
Thank you @ievacerny for putting this draft together. It will help discussion indeed.

> But if it's fine to attach a default method to all mapped traits, then the code could definitely be simplified. An issue with this would be attempting to map an undefined key trait (when it doesn't have a default value).

I don't know if I understand the "all mapped traits" part here. Do you mean all instances where `mapped_traits_for` is called?

> An issue with this would be attempting to map an undefined key trait (when it doesn't have a default value).

Perhaps the suggestion @mdickinson made in https://github.com/enthought/traits/issues/1081#issuecomment-628801064 to do: `if value is Undefined: return Undefined` could solve this?  The mapped trait's default method (like the one you defined) can do this `Undefined` check instead of deferring it to `mapped_value` (see below).

By reading `update_traits_class_dict`, I believe there are three ways to define a default:
(1) Statically, on the trait:
```
        class Person(HasTraits):
            married = Map({"yes": 1, "no": 0}, default_value="no")
```

(2) Dynamic default (the issue being fixed here):
```
        class Person(HasTraits):
            married = Map({"yes": 1, "no": 0})
            def _married_default(self):
                return "no"
```
(3) Overriding a trait with a constant in subclass:
```
        class Person(HasTraits):
            married = Map({"yes": 1, "no": 0})

        class SinglePerson(HasTraits):
            married = "no"
```

There are other occurrences of `mapped_traits_for`, to support `add_trait` and `add_class_trait`. From what I can tell, `add_trait` and `add_class_trait` do not support dynamic defaults (as `set_default_value` is not called with a `*_default` method there), so we don't need to worry about those (for #1081 at least).

Currently the `mapped_traits_for` update the default value for the mapped trait, e.g. `mapped_traits_for` is called again here for case (3):
https://github.com/enthought/traits/blob/be82bc039013e8d9beda29b4255458517f74bcc1/traits/has_traits.py#L506
So cases (1) and (3) are supported by the current definition of `mapped_traits_for` and code path in `update_traits_class_dict`.
The code here adds support for case (2)

The new `mapped_traits_for` I was thinking of looks like this:
```
from .trait_types import Any
def mapped_trait_for(trait, name):
    """ Returns the 'mapped trait' definition for a mapped trait.
    """
    mapped_trait = Any(
        is_base=False, transient=True, editable=False
    ).as_ctrait()

    def default(instance):
        value = getattr(instance, name, Undefined)
        if value is Undefined:
            return Undefined
        return trait.handler.mapped_value(value)

    mapped_trait.set_default_value(DefaultValue.callable, default)
    return mapped_trait
```

I believe this will support all cases (1) - (3), and we would not need this line for case (3):
https://github.com/enthought/traits/blob/be82bc039013e8d9beda29b4255458517f74bcc1/traits/has_traits.py#L506

I have only looked into `Map`, i.e. I have not looked into other trait types where `is_mapped` is True. If there are a fourth or a fifth way of defining default on the original trait, then this approach will also need to be reassessed.

Another thing worth checking is whether the static default on the mapped trait is relied on by a separate code path, e.g. this one:
https://github.com/enthought/traits/blob/be82bc039013e8d9beda29b4255458517f74bcc1/traits/traits.py#L288
https://github.com/enthought/traits/blob/be82bc039013e8d9beda29b4255458517f74bcc1/traits/traits.py#L395
So far it looks to me that the new `mapped_trait_for` would not change the existing code path in `_TraitMaker.define`.

**Comment by mdickinson** [comment]
One corner case we need to consider is where the static default is overridden by a method-based default in a subclass. Whatever solution we go with, I'd recommend adding a unit test for that case.

**Comment by ievacerny** [comment]
> > But if it's fine to attach a default method to all mapped traits, then the code could definitely be simplified. An issue with this would be attempting to map an undefined key trait (when it doesn't have a default value).
> 
> I don't know if I understand the "all mapped traits" part here. Do you mean all instances where `mapped_traits_for` is called?

With this approach mapped traits created using `mapped_traits_for` (which should be the case for all mapped traits that exist) would have a `DefaultValue.callable` as their default value type. With draft solution I was specifically targeting only case (2) so that the mapped trait would only have `DefaultValue.callable` _IF_ the key trait also has `DefaultValue.callable`. So more specifically what I was wondering in that comment was if it's okay to have different default types for key an mapped traits. I was working with the assumption that they should match but maybe that's not necessary. In that case your suggested solution would be a very nice way to address the problem!

**Comment by mdickinson** [comment]
> So more specifically what I was wondering in that comment was if it's okay to have different default types for key an mapped traits.

I was wondering the same thing: it seems cleaner _not_ to rely on callable defaults if we don't have to. I'm not sure I have a great reason for that, except possibly that things get messy when subclassing and overriding something with a callable default (we have other issues open about this).

**Comment by kitchoi** [comment]
I guess what I am not sure about is why the mapped trait's default should care how the default "key trait" (is it what this is called?) is defined. e.g. why should the mapped trait default be defined differently if the key trait's default is static/overridden/dynamic?

> possibly that things get messy when subclassing and overriding something with a callable default (we have other issues open about this).

I did not know that callable default can be overridden in subclass for mapped traits. Could you point me to an example please (maybe one these open issues)?  By "callable default", you meant a default where the `default_value_type` is `DefaultValue.callable`, right?

**Comment by kitchoi** [comment]
> callable default can be overridden in subclass for mapped traits.

Oh I think I got confused, I think you meant subclassing the trait type?  I was thinking about subclassing a subclass of HasTraits (too many subclasses).

**Comment by mdickinson** [comment]
So I meant situations like this:

```
class Person(HasTraits):
    married = Map({'yes': 1, 'no': 0, 'unknown': 2}, default_value="unknown")


class Child(Person):
    married = "no"


class Pet(Person):
    def _married_default(self):
        return "no"
```
... and other variations on the same theme.

Whatever changes we make, we need to make sure that `Child` and `Pet` continue to work at least as well as they did before (however well that may be).

My recollection is that this sort of subclassing doesn't work well for `CTrait` objects whose default_value_type is `DefaultValueType.callable`. But it may be fine.

So the message is just that there's a risk in changing the `default_value_type` of the shadow trait; we need to be sure that we don't accidentally break anything user-facing as a result of that change.

**Comment by kitchoi** [comment]
I see. Thank you. Yes I agree we will need tests for these various ways of defining/overriding defaults.

File: traits/has_traits.py (hunk 2 above)
Comment (ievacerny) [reply traits/has_traits.py]: No, that was just a remnant from when I was still playing with `lambda`. Pickling was working without it as well.

File: traits/has_traits.py (hunk 3 above)
Comment (ievacerny) [reply traits/has_traits.py]: > I guess the solution I was thinking of was different: that is, in the branch `if default is not None` below, set both the default value of the main trait _and_ (in the case of a mapped trait) the default value of the shadow trait at the same time. Would that work, or is there something I'm missing?

It would've worked and would've made the code a bit cleaner compared to the original approach in the draft. Not relevant to the new approach, so marking this as resolved.

**Comment by ievacerny** [comment]
Switched the PR to follow the approach suggested by @kitchoi -- thank you for your help! And thank you both @kitchoi and @mdickinson for a helpful discussion.

Also added more tests for setting and overriding defaults of `Map`, `PrefixMap` and `Expression` traits to verify that the bug has been fix and other user-facing behaviour has not changed.

File: traits/tests/test_expression.py [hunk 4]
```diff
@@ -53,3 +64,74 @@ def _bar_default(self):
         msg = "The 'bar' trait of a Foo instance must be a valid"
         with self.assertRaisesRegex(TraitError, msg):
             f.bar
+
+    @unittest.expectedFailure  # FIXME issue #1096
```
Comment (kitchoi) [inline traits/tests/test_expression.py:68]: It would be better to wrap the part of the code you expect to fail using `assertRaises` instead of wrapping the entire test with `expectedFailure`. Despite the name, this decorator does not differentiate test errors from test failures. Test errors are considered expected failures as well and the test can end up being unmaintained.

File: traits/trait_converters.py [hunk 5]
```diff
@@ -109,17 +109,23 @@ def trait_for(trait):
         return Trait(trait)
 
 
-def mapped_trait_for(trait):
+def mapped_trait_for(trait, name):
     """ Returns the 'mapped trait' definition for a mapped trait.
```
Comment (kitchoi) [inline traits/trait_converters.py:113]: It would be a good opportunity to improve the docstring here given that we know what `trait` and `name` are.

File: traits/tests/test_expression.py (hunk 4 above)
Comment (mdickinson) [reply traits/tests/test_expression.py:68]: I'm fine with an expected failure if there's a plan to resolve this in the near future (e.g., before the 6.1 release).

File: traits/trait_converters.py (hunk 5 above)
Comment (mdickinson) [reply traits/trait_converters.py:113]: This is a backwards compatibility break for anyone using `mapped_trait_for`. Please can you do a quick check that we don't have code outside traits using `mapped_trait_for`?

It's a bit of a grey area: I think it's okay to change this---it's reasonable to say that if you're importing from anything other than one of the `*.api` modules then you're on your own with respect to backwards compatibility breakage. But we should still make the check.

Also, agreed with @kitchoi that we should update the docstring. The traits stubs may also need updating.

**Comment by mdickinson** [comment]
So I'm not sure whether we have a pickling problem or not. The shadow trait is no longer pickleable:

```
>>> from traits.api import *
>>> import pickle
>>> class Person(HasTraits):
...     married = Map({"no":13, "yes":22, "itscomplicated": 4})
... 
>>> shadow_trait = Person().trait("married_")
>>> pickle.dumps(shadow_trait)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: Can't pickle local object 'mapped_trait_for.<locals>.default'
```

Same test on master:
```
>>> from traits.api import *
>>> import pickle
>>> class Person(HasTraits):
...     married = Map({"no":13, "yes":22, "itscomplicated": 4})
... 
>>> shadow_trait = Person().trait("married_")
>>> pickle.dumps(shadow_trait)
b'\x80\x04\x95)\x01\x00\x00\x00\x00\x00\x00\x8c\rtraits.ctrait\x94\x8c\x06CTrait\x94\x93\x94K\x00\x85\x94\x81\x94(K\x00K\x00K\x05NK\x03NK\x00\x8c\x11traits.trait_base\x94\x8c\n_Undefined\x94\x93\x94)R\x94K\x00NNK\x04N\x8c\x12traits.trait_types\x94\x8c\x03Any\x94\x93\x94)\x81\x94}\x94(\x8c\rdefault_value\x94h\x08\x8c\t_metadata\x94}\x94(\x8c\x07is_base\x94\x89\x8c\ttransient\x94\x88\x8c\x08editable\x94\x89\x8c\x04type\x94\x8c\x05trait\x94u\x8c\x12default_value_type\x94\x8c\x10traits.constants\x94\x8c\x0cDefaultValue\x94\x93\x94K\x00\x85\x94R\x94ub}\x94(h\x11\x89h\x12\x88h\x13\x89h\x14h\x15ut\x94b.'
```

This _might_ not matter: since the shadow trait is marked as transient, I guess it doesn't get pickled anyway when you pickle an object.

But if we can find an easy way to keep pickleability of the Trait, then that would be good.

**Comment by mdickinson** [comment]
> But if we can find an easy way to keep pickleability of the Trait, then that would be good.

The standard trick of moving the current `default` inner function to a module-level function, and using `DefaultValue.callable_and_args` to pass the necessary arguments should work. (Though I haven't tested.)

File: traits/tests/test_expression.py (hunk 4 above)
Comment (mdickinson) [reply traits/tests/test_expression.py:68]: An alternative would be to remove these tests from this PR and attach them to the corresponding issue instead. Then the PR that fixes the issue can re-instate them.

File: traits/tests/test_expression.py [hunk 6]
```diff
@@ -53,3 +64,74 @@ def _bar_default(self):
...
+        class BaseFoo(HasTraits):
+            # The default value set in the class definition is "0"
+            bar = Expression()
+
+        class Foo(BaseFoo):
+            bar = "3"
+
+        f = Foo()
+        self.assertEqual(f.bar, "3")
+        self.assertEqual(f.bar_, compile("3", "<string>", "eval"))
```
Comment (mdickinson) [inline traits/tests/test_expression.py:79]: I'm quite surprised that it's possible to compare code objects for equality this way. If we didn't want to rely on this, it could make sense to test for behaviour instead of exact equality. (i.e., we don't much care what `f.bar_` actually is; we only care that we can `eval` it in a suitable context and get the expected results).

Comment (ievacerny) [reply traits/tests/test_expression.py:79]: Replaced

File: traits/trait_converters.py (hunk 5 above)
Comment (ievacerny) [reply traits/trait_converters.py]: Couldn't find any other usage of `mapped_trait_for` on Github. 

Updated the docstring. `mapped_trait_for` signature isn't in traits stubs so no update there.

File: traits/tests/test_expression.py (hunk 4 above)
Comment (ievacerny) [reply traits/tests/test_expression.py]: Wrapped the tests in `assertRaises` because the fix to the bug is not as straight forwards as expected.

**Comment by ievacerny** [comment]
> > But if we can find an easy way to keep pickleability of the Trait, then that would be good.
> 
> The standard trick of moving the current `default` inner function to a module-level function, and using `DefaultValue.callable_and_args` to pass the necessary arguments should work. (Though I haven't tested.)

I don't think it's possible to pass the `HasTraits` object via `DefaultValue.callable_and_args` but I managed to get a pickleable shadow trait using `partial`.

I hadn't realised that my "pickle" test wasn't even attempting to pickle the shadow trait, so I also added more tests that actually do that.

**Comment by mdickinson** [comment]
> I don't think it's possible to pass the `HasTraits` object via `DefaultValue.callable_and_args`

Agreed; it looks that way. :-(  Using `partial` sounds like a reasonable alternative.

File: traits/trait_converters.py [hunk 7]
```diff
@@ -109,17 +109,43 @@ def trait_for(trait):
...
+            The name of the trait for which the 'mapped trait' definition is
+            being created.
+
+        Returns
+        -------
+        trait_types.Any
+            A definition of the 'mapped trait'
     """
-    from .trait_types import Any
+    from functools import partial
```
Comment (mdickinson) [inline traits/trait_converters.py:136]: I think we can lift this import to module-level; no need to make it local to the function.

File: traits/trait_converters.py [hunk 8]
```diff
@@ -109,17 +109,43 @@ def trait_for(trait):
...
 
-    default_value = trait.default_value()[1]
-    try:
-        default_value = trait.handler.mapped_value(default_value)
-    except Exception:
-        pass
+    from .constants import DefaultValue
```
Comment (mdickinson) [inline traits/trait_converters.py:138]: Do these imports need to be inside the function? (I could believe that there might be a circular import issue with the `Any` import, but I think the other two should be okay.)

Comment (ievacerny) [reply traits/trait_converters.py:138]: No, `DefaultValue` and `Undefined` don't have to. Moving them out
